File: sections/14_core_patterns_mixed_contest/problems/i_wildcard_parentheses/solution.py
import sys


def possible(s):
    if len(s) % 2:
        return False
    low = high = 0
    for index, char in enumerate(s):
        if char == 40:
            low += 1
            high += 1
        elif char == 41:
            low -= 1
            high -= 1
        else:
            low -= 1
            high += 1
        if low < 0:
            low = (index + 1) & 1
        if high < 0:
            return False
    return low == 0


def main() -> None:
    data = sys.stdin.buffer.read().split()
    print("\n".join("YES" if possible(s) else "NO" for s in data[1:]))


if __name__ == "__main__":
    main()

File: sections/14_core_patterns_mixed_contest/problems/i_wildcard_parentheses/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int tests;
    cin >> tests;
    while (tests--) {
        string s;
        cin >> s;
        int low = 0, high = 0;
        bool possible = s.size() % 2 == 0;
        for (int index = 0; index < (int)s.size(); ++index) {
            char c = s[index];
            if (c == '(') ++low, ++high;
            else if (c == ')') --low, --high;
            else --low, ++high;
            if (low < 0) low = (index + 1) & 1;
            if (high < 0) possible = false;
        }
        cout << (possible && low == 0 ? "YES" : "NO") << '\n';
    }
}
